### PR TITLE
PM-4109 Vault Onboarding M2

### PR DESCRIPTION
**Pull request by Jingo88**
```
- [ ] Bug fix
- [ X ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

- Users Web Vault will check if the Bitwarden Extension is installed and will complete the task in the onboarding component
- When all tasks completed the onboarding component will remain on the page until refresh/log out +log in
- Once the component is dismissed or all tasks are completed. Onboarding will not show again even if the user deletes extension, or deletes all items in vault
- The text for the Import Data field has been updated. The second sentence is dynamic based on if user is confirmed in orgs
- Install Link is now an `<a>` tag for A11y
- All features from Onboarding M1 remain the same

NOTE: The screen recording below will cover these use cases

## Code changes

- `content-message-handler` - add a listener that will post a message saying the extension does exist
- `vault-onboarding.component` 
   - check for organizations for dynamic text in import data. 
   - Made some logic updates to keep the onboarding component on screen while all tasks are completed, until the user refreshes or logs out
   - Added methods to post and listen for if extension exists
- `messages.json` updated text fields and descriptions
- spec files updated

## Screen Recording

https://github.com/bitwarden/clients/assets/8302660/90047a43-3e28-4e04-bdef-43f60a25037e

### CODE REVIEW

File: apps/browser/src/autofill/content/content-message-handler.ts [hunk 1]
```diff
@@ -24,6 +24,7 @@ const windowMessageHandlers: ContentMessageWindowEventHandlers = {
     handleAuthResultMessage(data, referrer),
   webAuthnResult: ({ data, referrer }: { data: any; referrer: string }) =>
     handleWebAuthnResultMessage(data, referrer),
+  checkIfBWExtensionInstalled: () => window.postMessage({ command: "hasBWInstalled" }),
```
Comment (cagonzalezcs) [inline apps/browser/src/autofill/content/content-message-handler.ts:27]: Looks like we're missing a couple of things here.

- The typing information for this key hasn't been added to the `ContentMessageWindowEventHandlers` type. Minor nit, but it wouldn't be a bad idea to facilitate inclusion of that typing information
- This is a nit as well, but would it be possible to pull the `window.postMessage({ command: "hasBWInstalled" })` to a separate method to keep consistent with the way these handlers are called?
- We seem to be missing a test for this added handler

File: apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.html [hunk 2]
```diff
@@ -27,14 +27,21 @@
         <button type="button" bitLink (click)="emitToAddCipher()">
           {{ "onboardingImportDataDetailsLink" | i18n }}
         </button>
-        {{ "onboardingImportDataDetailsPartTwo" | i18n }}
+        <span *ngIf="orgs.length === 0">
```
Comment (shane-melton) [inline apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.html:30]: question: Any concern for a `null` orgs array here? We always seem to have issues with org arrays being null on startup and throwing exceptions.

File: apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts [hunk 3]
```diff
@@ -84,6 +89,25 @@ export class VaultOnboardingComponent implements OnInit, OnChanges, OnDestroy {
     this.destroy$.complete();
   }
 
+  checkForBrowserExtension() {
+    if (this.showOnboarding) {
+      window.postMessage({ command: "checkIfBWExtensionInstalled" });
+      window.addEventListener("message", (event) => this.getMessages(event));
```
Comment (shane-melton) [inline apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts:95]: issue: We need to make sure we remove this `eventListener` when this component is destroyed. At the moment, this appears to re-add a new listener every time the vault onboarding component is rendered. 

Ideally, we should be able to make use of the [`fromEvent`](https://rxjs.dev/api/index/function/fromEvent) rxjs function and subscribe with our normal `takeUntil(destroy$)` pattern so we don't have to worry about explicitly cleaning up the event listener.

**Review by shane-melton — CHANGES_REQUESTED** [review]
We need to make sure we clean up any window event listeners we add. Otherwise, this looks good! Nice work!

File: apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts (hunk 3 above)
Comment (cagonzalezcs) [reply apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts:95]: Also keep in mind that if you pass an anonymous function to a handler, there is no way to clean up the listener's associate with that handler.

You need to ensure you name the method in order to properly call `window.removeEventListener` on it.

File: apps/browser/src/autofill/content/content-message-handler.ts (hunk 1 above)
Comment (Jingo88) [reply apps/browser/src/autofill/content/content-message-handler.ts]: thanks. cleaned up the typings and made a separate method

File: apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.html (hunk 2 above)
Comment (Jingo88) [reply apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.html]: hmmmm good call. So far I haven't run into anything other than `[]` but i'll put a null check in there for safety

File: apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts (hunk 3 above)
Comment (Jingo88) [reply apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts]: awesome. appreciate it

**Review by shane-melton — CHANGES_REQUESTED** [review]
Looking good! Just one more question and small issue with the null check.

File: apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts [hunk 4]
```diff
@@ -92,7 +90,11 @@ export class VaultOnboardingComponent implements OnInit, OnChanges, OnDestroy {
   checkForBrowserExtension() {
     if (this.showOnboarding) {
       window.postMessage({ command: "checkIfBWExtensionInstalled" });
-      window.addEventListener("message", (event) => this.getMessages(event));
+      fromEvent<MessageEvent>(window, "message")
```
[the remark this answers is not in the document]
Comment (Jingo88) [reply apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts:92]: ah good calls. moved the code around. thanks

**Review by shane-melton — APPROVED** [review]

**Review by cagonzalezcs — APPROVED** [review]

**Review by cagonzalezcs — APPROVED** [review]

**Review by shane-melton — APPROVED** [review]

**Pull request merged**